Sources/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -610,25 +627,38 @@ extension PurchasesOrchestrator: StoreKit2StorefrontListenerDelegate {
 
 }
 
-// MARK: Private funcs.
+// MARK: Private funcs
+
 private extension PurchasesOrchestrator {
 
-    func getAndRemovePurchaseCompletedCallback(
-        forTransaction transaction: SKPaymentTransaction
-    ) -> PurchaseCompletedBlock? {
-        guard let productIdentifier = transaction.productIdentifier else {
-            return nil
+    func addPurchaseCompletedCallback(
+        productIdentifier: String,
+        completion: @escaping PurchaseCompletedBlock
+    ) {
+        self.purchaseCompleteCallbacksByProductID.modify { callbacks in
+            guard callbacks[productIdentifier] == nil else {
+                completion(nil, nil, ErrorUtils.operationAlreadyInProgressError(), false)
+                return
+            }
+            callbacks[productIdentifier] = completion
         }
+    }
 
+    func getAndRemovePurchaseCompletedCallback(
+        forTransaction transaction: StoreTransaction
+    ) -> PurchaseCompletedBlock? {
         return self.purchaseCompleteCallbacksByProductID.modify {
-            $0.removeValue(forKey: productIdentifier)
+            $0.removeValue(forKey: transaction.productIdentifier)
         }
     }
 
-    func fetchProductsAndPostReceipt(withTransaction transaction: SKPaymentTransaction, receiptData: Data) {
-        if let productIdentifier = transaction.productIdentifier {
+    func fetchProductsAndPostReceipt(
+        withTransaction transaction: StoreTransaction,
+        receiptData: Data,
+        storefront: StorefrontType?
+    ) {
+        if let productIdentifier = transaction.productIdentifier.notEmpty {
             self.products(withIdentifiers: [productIdentifier]) { products in
-                let storefront = self.storeKitWrapper.currentStorefront
```

Comment:
If the `productIdentifier` is empty, it means we got a bad `SKTransaction`. 
Current behavior is to just not send it to the backend. 

But... 
At this point, the user got charged. They won't have access, and they'll get an error. That's a terrible experience. Sure, it's because of a bug in StoreKit, but it's a bug in the price, and the only value of fetching the product here is to send the prices to our backend. So the user gets a bad experience but we're not preventing any issues on their side, this is just for accounting purposes. 

The user should come first. If they made a purchase, and it went through, we should prioritize granting them access. Getting the products comes second. 

This is illustrated in the current flow by the fact that we don't check whether we actually could fetch the product - there's no guarantee that we will (the request could fail) but we don't care, we just send the receipt. 

We should do the same for the case where the SKTransaction is broken. The receipt should still be fine, so even if the transaction is broken, we can send the transaction to the backend, and get the customerInfo, and preserve user experience, in exchange for slightly worse analytics when `SKTransaction`s are broken. 

It's not like the customer would be able to fix this by retrying either, they'll already have purchased the product

Reply:
So I'd suggest to: 
- Log when the transaction is bad, that'll help us debug issues with prices
- Always refresh and send the receipt, even if the transaction was bad

Reply:
For context, this PR doesn't change the existing behavior. The reason I changed this to `notEmpty` is because of [this](https://github.com/RevenueCat/purchases-ios/blob/4.5.2/Sources/Purchasing/StoreKitAbstractions/SK1StoreTransaction.swift#L21) (`StoreTransactionType.productIdentifier` is not `Optional`) whereas previously the `nil` check did that.

> Log when the transaction is bad, that'll help us debug issues with prices

That's always done by [`SKPaymentTransaction.productIdentifier`](https://github.com/RevenueCat/purchases-ios/blob/4.5.2/Sources/Purchasing/StoreKitAbstractions/SK1StoreTransaction.swift#L43).

> Always refresh and send the receipt, even if the transaction was bad

I think that makes sense. So to make sure I understand, basically change the `else` to:
```swift
self.postReceipt(withTransaction: transaction,
                 receiptData: receiptData,
                 products: [], // No product data, but still post the receipt
                 storefront: storefront)
```

I can add unit tests for this too.

Reply:
Kinda scary that this doesn't make any test fail:
<img width="499" alt="Screen Shot 2022-06-09 at 11 30 47" src="https://user-images.githubusercontent.com/685609/172919172-351b759a-dfba-48a9-a032-d3b9dc378ef7.png">

Reply:
Huh I see, there's just no test coverage for it.
It is an edge case after all, but I'll add a test.

Reply:
Turns out this breaks during tests, because we rely on product identifiers for the callbacks keys...

Reply:
I'm gonna leave this for a separate PR to deal with this problem.

Reply:
Filed [CSDK-119]

Reply:
I also added a failing test: https://github.com/RevenueCat/purchases-ios/pull/1680